File: apps/demo/modules/expo-module-showcase/android/src/main/java/expo/modules/showcase/ShowcaseObjects.kt
package expo.modules.showcase

import expo.modules.kotlin.AppContext
import expo.modules.kotlin.sharedobjects.SharedObject
import expo.modules.kotlin.sharedobjects.SharedRef

class ShowcaseSharedCounter(appContext: AppContext, @Volatile var value: Double) : SharedObject(appContext) {
  @Synchronized fun increment(delta: Double): Double {
    value += delta
    return value
  }
}

class ShowcaseTextRef(appContext: AppContext, value: String) : SharedRef<String>(value, appContext) {
  override val nativeRefType = "expo-module-showcase.text"
}

File: apps/demo/modules/expo-module-showcase/ios/ShowcaseObjects.swift
import ExpoModulesCore
import Foundation

final class ShowcaseSharedCounter: SharedObject {
  private let lock = NSLock()
  private var storedValue: Double
  var value: Double {
    get { lock.lock(); defer { lock.unlock() }; return storedValue }
    set { lock.lock(); defer { lock.unlock() }; storedValue = newValue }
  }

  init(_ value: Double) {
    storedValue = value
    super.init()
  }

  func increment(_ delta: Double) -> Double {
    lock.lock()
    defer { lock.unlock() }
    storedValue += delta
    return storedValue
  }
}

final class ShowcaseTextRef: SharedRef<String> {
  override var nativeRefType: String { "expo-module-showcase.text" }
}
